sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAnalyticFunctions.java

```diff
@@ -42,4 +44,77 @@
     throw new UnsupportedOperationException(
         String.format("Analytics Function [%s] is not supported", functionName));
   }
+
+  public static <T> Combine.CombineFn<T, ?, T> navigationFirstValue() {
+    return new FirstValueCombineFn();
+  }
+
+  public static <T> Combine.CombineFn<T, ?, T> navigationLastValue() {
+    return new LastValueCombineFn();
+  }
+
+  private static class FirstValueCombineFn<T> extends Combine.CombineFn<T, Optional<T>, T> {
+    private FirstValueCombineFn() {}
+
+    @Override
+    public Optional<T> createAccumulator() {
+      return Optional.empty();
+    }
+
+    @Override
+    public Optional<T> addInput(Optional<T> accumulator, T input) {
+      Optional<T> r = accumulator;
+      if (!accumulator.isPresent()) {
+        r = Optional.of(input);
+      }
+      return r;
+    }
+
+    @Override
+    public Optional<T> mergeAccumulators(Iterable<Optional<T>> accumulators) {
+      Optional<T> r = Optional.empty();
+      for (Optional<T> ac : accumulators) {
+        if (!r.isPresent() && ac.isPresent()) {
+          r = ac;
```

Comment:
you can optimize this for loop by adding a break, because `r.isPresent()` will be always true after first `r=ac`.

Reply:
Since, navigation functions rely on a sorted input and the addInput method actually should be called synchronously, the mergeAccumulators method never is invoked. So, I was wondering If I would rather remove this code and leave a mock implementation (let's say return null). What do you think?

Reply:
Ah I see. Yes it's actually called sequentially on the sorted input.

I think you can through an [UnsupportedOperationException](https://docs.oracle.com/javase/7/docs/api/java/lang/UnsupportedOperationException.html) since you will not call mergeAccumulators.

And in the future if there is usage to call `mergeAccumulators`, the exception will be hit and people will know they need to implement it.